homeassistant/components/saj/coordinator.py

```diff
@@ -0,0 +1,91 @@
+"""Coordinator to fetch data from SAJ solar inverter."""
+import asyncio
+import logging
+from typing import Callable
+
+import pysaj
+
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DEFAULT_NAME, UPDATE_INTERVAL
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _init_pysaj(wifi, host, username, password):  # pragma: no cover
+    kwargs = {"wifi": wifi}
+    if username and password:
+        kwargs["username"] = username
+        kwargs["password"] = password
+
+    return pysaj.SAJ(host, **kwargs)
+
+
+class SAJInverter(DataUpdateCoordinator):
+    """Representation of a SAJ inverter."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        name="",
+        wifi=True,
+        host=None,
+        username=None,
+        password=None,
+        saj=None,
+    ):
+        """Init SAJ Inverter class."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=name or DEFAULT_NAME,
+            update_interval=UPDATE_INTERVAL,
+            update_method=self.update,
+        )
+        self.last_update_success = False
+        self._saj = saj or _init_pysaj(wifi, host, username, password)
+        self._sensor_def = pysaj.Sensors(wifi)
+
+    def get_enabled_sensors(self):
+        """Return enabled sensors keys."""
+        return [s.key for s in self._sensor_def if s.enabled]
+
+    @property
+    def serialnumber(self):
+        """Return the serial number of the inverter."""
+        return self._saj.serialnumber
+
+    async def connect(self):
+        """Try to connect to the inverter."""
+        try:
+            done = await self._saj.read(self._sensor_def)
+            if done:
+                return
+        except asyncio.TimeoutError:
+            pass
+
+        raise CannotConnect
+
+    async def setup(self, async_add_entities: Callable):
+        """Add sensors to Core and get first state."""
+        # pylint: disable=import-outside-toplevel
+        from .sensor import SAJSensor
```

Comment:
uh? Why do we need to import outside of the top level?

Reply:
Was a circular dependency error here. Coordinator vs the sensor. Any suggestion?

Reply:
Yes, move it into the sensor platform :)